Make roll a decomposition and remove its lowering

lezcano opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #119884
* __->__ #119857
* #119864
* #119863

We use the fact that we now propagate indexing properly to avoid having
to maintain two different implementations of the op. Doing this we also remove
a spurious guard on this op.

We move the ref into a decomp as we now use advanced indexing.
The only difference we did in the implementation is that we now use
advanced indexing rather than `torch.cat`.

We also remove it from core. Let's see how this goes.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @amjames

lezcano committed:
Make roll a decomposition and remove its lowering

We use the fact that we now propagate indexing properly to avoid having
to maintain two different implementations of the op.

We move the ref into a decomp as we now use advanced indexing.
The only difference we did in the implementation is that we now use
advanced indexing rather than `torch.cat`.

We also remove it from core. Let's see how this goes.

[ghstack-poisoned]

lezcano committed:
Update on "Make roll a decomposition and remove its lowering"


We use the fact that we now propagate indexing properly to avoid having
to maintain two different implementations of the op. Doing this we also remove
a spurious guard on this op.

We move the ref into a decomp as we now use advanced indexing.
The only difference we did in the implementation is that we now use
advanced indexing rather than `torch.cat`.

We also remove it from core. Let's see how this goes.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

lezcano committed:
Update on "Make roll a decomposition and remove its lowering"


We use the fact that we now propagate indexing properly to avoid having
to maintain two different implementations of the op. Doing this we also remove
a spurious guard on this op.

We move the ref into a decomp as we now use advanced indexing.
The only difference we did in the implementation is that we now use
advanced indexing rather than `torch.cat`.

We also remove it from core. Let's see how this goes.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

lezcano commented:
cc @SherlockNoMad @larryliu0820 regarding the core removal

lezcano committed:
Update on "Make roll a decomposition and remove its lowering"


We use the fact that we now propagate indexing properly to avoid having
to maintain two different implementations of the op. Doing this we also remove
a spurious guard on this op.

We move the ref into a decomp as we now use advanced indexing.
The only difference we did in the implementation is that we now use
advanced indexing rather than `torch.cat`.

We also remove it from core. Let's see how this goes.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

peterbell10 commented on torch/_decomp/__init__.py:
```diff
@@ -385,6 +385,7 @@ def core_aten_decompositions() -> Dict[torch._ops.OperatorBase, Callable]:
             aten.renorm,
             aten.renorm_,
             aten.replication_pad2d,
+            aten.roll,
```
Should this be inductor-only to not incur the revert from export.

lezcano replied:
That's the thing. I think this should be a core decomposition. As the lowering implies, this was not a core decomp from the get go just because we were missing the constant propagation pass that you implemented.

peterbell10 replied:
I'm not comfortable approving without resolving this.

peterbell10 replied:
I'm not comfortable approving without resolving this.

peterbell10 commented on torch/_refs/__init__.py:
```diff
@@ -3738,9 +3738,8 @@ def roll(
-    t1 = torch.narrow(a, dim, 0, start)
-    return torch.cat((t0, t1), dim)
+    idx = torch.arange(size, device=a.device)
+    return a.index_select(dim, torch.fmod(start + idx, size))
```
Nice. I take it the codegen is the same?

lezcano replied:
I tested on the OpInfos running it with `PYTORCH_ALL_SAMPLES=1 TORCH_LOGS_FORMAT="%(message)s" TORCH_LOGS=output_code pytest test/inductor/test_torchinductor_opinfo.py -vsk test_comprehensive_roll_cuda_int64` and it does.

lezcano committed:
Update on "Make roll a decomposition and remove its lowering"


We use the fact that we now propagate indexing properly to avoid having
to maintain two different implementations of the op. Doing this we also remove
a spurious guard on this op.

We move the ref into a decomp as we now use advanced indexing.
The only difference we did in the implementation is that we now use
advanced indexing rather than `torch.cat`.

We also remove it from core. Let's see how this goes.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

lezcano committed:
Update on "Make roll a decomposition and remove its lowering"


We use the fact that we now propagate indexing properly to avoid having
to maintain two different implementations of the op. Doing this we also remove
a spurious guard on this op.

We move the ref into a decomp as we now use advanced indexing.
The only difference we did in the implementation is that we now use
advanced indexing rather than `torch.cat`.

We also remove it from core. Let's see how this goes.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

lezcano commented on torch/testing/_internal/opinfo/definitions/fft.py:
```diff
@@ -823,10 +823,18 @@ def mt(shape, **kwargs):
         "_refs.fft.ifftshift",
         op_db=op_db,
         torch_opinfo_name="fft.ifftshift",
+        skips=(
+            # TODO Move ifftshift to decomps
+            DecorateInfo(unittest.expectedFailure, 'TestCommon', 'test_python_ref'),
+        ),
```
@peterbell10 I had to add these. Do you want me to append a PR moving these ops to decomps and removing these xfails, or do you prefer to leave them like this?

peterbell10 replied:
Seems fine to keep it as a skip imo

lezcano committed:
Update on "Make roll a decomposition and remove its lowering"


We use the fact that we now propagate indexing properly to avoid having
to maintain two different implementations of the op. Doing this we also remove
a spurious guard on this op.

We move the ref into a decomp as we now use advanced indexing.
The only difference we did in the implementation is that we now use
advanced indexing rather than `torch.cat`.

We also remove it from core. Let's see how this goes.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

lezcano commented:
let's wait for the Export team, see what they say.

larryliu0820 commented:
@lezcano Thanks for tagging us!

I think from BC/FC perspective this is ok, since we are decomposing a core ATen op into other core ATen ops, there shouldn't be any BC/FC breakage. In ExecuTorch we plan to implement all core ATen ops in a portable way (cc @manuelcandales ) so maybe we don't have to implement `roll` anymore.

Tagging others to double check if this is ok: @angelayi @SS-JIA @SherlockNoMad

angelayi commented:
I think it's fine? Checked internally and there doesn't seem to be any uses or special casing on this op. Will let @SherlockNoMad confirm

lezcano commented:
@SherlockNoMad PTAL

lezcano commented:
@pytorchbot merge
